File: virtual-and-abstract-AdamWalters1/notes/override_ex.cpp
#include <iostream>
using namespace std;

// Base class
class Animal {
public:
    // Virtual function
    virtual void speak() const {
        cout << "This animal does not have a sound" << endl;
    }
    // Virtual destructor to ensure proper cleanup of derived objects
    virtual ~Animal() {}
};

// Derived class
class Dog : public Animal {
public:
    // Override speak function
    void speak() const override {
        cout << "Woof!" << endl;
    }
};

// Another derived class
class Cat : public Animal {
public:
    // Override speak function
    void speak() const override {
        cout << "Meow!" << endl;
    }
};

// Function to demonstrate polymorphism
void animalSound(const Animal& animal) {
    animal.speak();
}

int main() {
    Animal myAnimal;
    Dog myDog;
    Cat myCat;

    animalSound(myAnimal); // Outputs: This animal does not have a sound
    animalSound(myDog);    // Outputs: Woof!
    animalSound(myCat);    // Outputs: Meow!

    return 0;
}

File: virtual-and-abstract-AdamWalters1/notes/override_ex.py
class Animal:
    def speak(self):
        print("This animal does not have a sound")

class Dog(Animal):
    def speak(self):
        print("Woof!")

class Cat(Animal):
    def speak(self):
        print("Meow!")

def animal_sound(animal):
    animal.speak()

if __name__ == "__main__":
    myAnimal = Animal()
    myDog = Dog()
    myCat = Cat()

    animal_sound(myAnimal)  # Outputs: This animal does not have a sound
    animal_sound(myDog)     # Outputs: Woof!
    animal_sound(myCat)     # Outputs: Meow!
